src/storage/src/render/sources.rs

```diff
@@ -466,10 +471,26 @@ fn append_metadata_to_value<G: Scope>(
 fn upsert_commands<G: Scope>(
     input: Collection<G, DecodeResult, Diff>,
     upsert_envelope: UpsertEnvelope,
-) -> Collection<G, (UpsertKey, Option<Result<Row, UpsertError>>, MzOffset), Diff> {
+    metadata_arity: usize,
+) -> Collection<G, (UpsertKey, Option<Result<Row, UpsertError>>, UpsertOrder), Diff> {
+    // Converting from source indices to metadata indices for optional order by
+    let key_value_arity = upsert_envelope.source_arity - metadata_arity;
+    let metadata_indices = upsert_envelope
+        .order_by_indices
+        .iter()
+        .map(|row_pos| row_pos - key_value_arity)
+        .collect();
+
     let mut row_buf = Row::default();
     input.map(move |result| {
-        let order = result.position;
+        let metadata = result.metadata;
+
+        let order = if upsert_envelope.order_by_indices.is_empty() {
+            result.position.into()
+        } else {
+            // Getting the ordering information from metadata
+            crate::render::upsert::get_order(&Ok(metadata.clone()), &metadata_indices)
```

Comment:
There is a bit of subtlety here: the `metadata` offset is guaranteed to be the same as `result.position` above! Might be worth a comment!

Reply:
True. Let me add the comment.

Reply:
Done